### PR TITLE
Review types for editing and navigation components

### PR DESCRIPTION
**Pull request by IlyaKhD**

### CODE REVIEW

File: js/ui/list.d.ts [hunk 1]
```diff
@@ -31,6 +37,87 @@ import {
...
+/** @public */
+export type ItemDeletingEvent = ComponentEvent<dxList> & ListItemInfo & {
+    cancel?: boolean | TPromise<void>;
+}
+
+/** @public */
+export type ItemHoldEvent = ComponentNativeEvent<dxList> & ListItemInfo;
+
+/** @public */
+export type ItemRenderedEvent = ComponentNativeEvent<dxList> & ItemInfo;
```
Comment (IlyaKhD) [inline js/ui/list.d.ts:87]: @DokaRus. other list item events have  *grouped* `itemIndex` that can be `number | { group: number; item: number; }`, but [ItemRendered](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxList/Configuration/#onItemRendered) declares `number` type only.
Is it correct?

**Comment by IlyaKhD** [comment]
@DokaRus, just to make sure: some components like [dxColorBox](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxColorBox/Events/) or [dxDropDownBox](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxDropDownBox/Events/) have no `contentReady` event, right?

File: js/events/index.d.ts [hunk 2]
```diff
@@ -6,6 +10,50 @@ export interface EventType { }
...
+    value?: any;
+}
+
+export interface Cancelable {
+    cancel?: boolean;
+}
+
+export interface ItemInfo {
+    readonly itemData?: any;
+    itemElement: TElement;
```
Comment (IlyaKhD) [inline js/events/index.d.ts]: `ItemInfo` interface is used in most of `ItemXXX` events. Is `.itemElement` field is always writable?

/cc @DokaRus, @IgnatovDan

File: js/events/index.d.ts [hunk 3]
```diff
@@ -6,6 +10,50 @@ export interface EventType { }
  */
 export type TEvent = {} extends EventType ? dxEvent : EventType;
 
+/** @public */
+export interface ComponentEvent<T> {
```
Comment (IlyaKhD) [inline js/events/index.d.ts]: These events are the base part for almost all events. Could you check if `readonly` modifier is applicable?
- `ComponentEvent`
- `ComponentNativeEvent`
- `ComponentDisposingEvent`
- `ComponentInitializedEvent`
- `ComponentOptionChangedEvent`

/cc @DokaRus, @IgnatovDan

File: js/events/index.d.ts [hunk 4]
```diff
@@ -6,6 +10,50 @@ export interface EventType { }
...
+    element: TElement;
+    model?: any;
+}
+
+/** @public */
+export interface ComponentNativeEvent<T> {
+    component: T;
+    element: TElement;
+    model?: any;
+    event?: TEvent;
```
Comment (IlyaKhD) [inline js/events/index.d.ts]: About this optional modifier - is it correct? The `event` field can be `undefined` even if an event is a *native-like* one?

/cc @DokaRus, @IgnatovDan

File: js/ui/tree_view.d.ts [hunk 5]
```diff
@@ -26,6 +30,80 @@ import {
...
+export type DisposingEvent = ComponentDisposingEvent<dxTreeView>;
+
+/** @public */
+export type InitializedEvent = ComponentInitializedEvent<dxTreeView>;
+
+/** @public */
+export type ItemClickEvent = ComponentNativeEvent<dxTreeView> & {
+    readonly itemData?: any;
+    readonly itemElement?: TElement;
+    readonly itemIndex?: number | any;
```
Comment (IlyaKhD) [inline js/ui/tree_view.d.ts:45]: @IgnatovDan, could we make this type more specific? [Documentation says](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxTreeView/Configuration/#onItemClick) it's `number | object` - does this `object` has a structure, like [dxList item events](https://github.com/DevExpress/DevExtreme/pull/16933/files#diff-6f112a3ec2f1220c5120dec4dd085da62d370c4f15044f761ee28fe915548361R43)?

Comment (IlyaKhD) [reply js/ui/tree_view.d.ts:45]: Same for the `ItemContextMenuEvent`

File: js/ui/form.d.ts [hunk 6]
```diff
@@ -40,6 +47,44 @@ import Widget, {
...
+}
+
+/** @public */
+export type InitializedEvent = ComponentInitializedEvent<dxForm>;
+
+/** @public */
+export type OptionChangedEvent = ComponentOptionChangedEvent<dxForm>;
+
+/** @public */
+export type GroupItemTemplateData = {
```
Comment (IlyaKhD) [inline js/ui/form.d.ts:73]: @IgnatovDan, could you review `GroupItemTemplateData` and `SimpleItemTemplateData`? I mean readonly and optional modifiers

File: js/events/index.d.ts (hunk 4 above)
Comment (IgnatovDan) [reply js/events/index.d.ts]: There is a code for the "args.event is null or undefined" config:
https://github.com/DevExpress/DevExtreme/blob/21_1/js/ui/context_menu/ui.context_menu.js#L861
https://github.com/DevExpress/DevExtreme/blob/21_1/js/ui/pivot_grid/ui.pivot_grid.field_chooser.js#L333

And, we have not checked the 'args.event' value on our tests.

imho we cannot declare it as 'not null'

File: js/events/index.d.ts (hunk 2 above)
Comment (IgnatovDan) [reply js/events/index.d.ts]: "itemElement" is readonly in https://github.com/IlyaKhD/DevExtreme/pull/1/files.

I checked my sources and found this declaration for TabPanel:
    readonly itemElement?: TElement;

and this declaration for TreeView:
    readonly itemElement?: TElement;
    readonly itemIndex?: number;

and this declaration for Accordion:
    readonly itemElement: TElement;

imho we can add 'readonly' here

File: js/ui/form.d.ts (hunk 6 above)
Comment (IgnatovDan) [reply js/ui/form.d.ts:73]: GroupItemTemplateData in https://github.com/IlyaKhD/DevExtreme/pull/1/files:

"component" is not null and it is readonly
"formData" can be null/undefined and it is readonly

https://github.com/DevExpress/DevExtreme/blob/21_1/js/ui/form/ui.form.js#L567
https://js.devexpress.com/Documentation/21_1/ApiReference/UI_Components/dxForm/Item_Types/GroupItem/#template

imho we can change to 'readonly component: dxForm;' + 'readonly formData?: any;'

Comment (IgnatovDan) [reply js/ui/form.d.ts:73]: SimpleItemTemplateData:

"component" is not null, readonly
other fields can be null/undefined, readonly

https://github.com/DevExpress/DevExtreme/blob/21_1/js/ui/form/ui.form.layout_manager.js#L924 - too complex code to review each property and no tests.
https://js.devexpress.com/Documentation/21_1/ApiReference/UI_Components/dxForm/Item_Types/SimpleItem/#template

imho we can change to 'readonly component: dxForm;' and 'readonly' for the rest members

File: js/ui/tree_view.d.ts (hunk 5 above)
Comment (IgnatovDan) [reply js/ui/tree_view.d.ts:45]: Yes, I saw it and wondered why there is 'object' but kept it 'as in docs'.
It seems it's time to dig deeper.

ItemClickEvent/ItemContextMenuEvent:

https://js.devexpress.com/Documentation/21_1/ApiReference/UI_Components/dxTreeView/Configuration/#onItemClick - "itemIndex" is described as "Number | Object"

Generated from declarations: https://github.com/DevExpress/DevExtreme/blob/21_1/js/ui/tree_view.d.ts#L109
Declarations were added here: https://github.com/DevExpress/DevExtreme/pull/5852/files#diff-0ad220537b603d00bc41a3c7617773006bfc16f6c9214778b647f62f7b2c8186R216 - no comments about "why and when an object can be passed in the index member"

https://js.devexpress.com/Documentation/21_1/ApiReference/UI_Components/dxList/Configuration/#onItemClick - similar "Object" but with a clear description WTF.

I have not found a TreeView code that controls this property.
IMHO its value is initialized here: https://github.com/DevExpress/DevExtreme/blob/21_1/js/ui/collection/ui.collection_widget.base.js#L1028

It seems to be an 'integer' (https://api.jquery.com/index/) but I cannot prove it (and, there is a similar function in our 'nonJQ' code).
Also, I even cannot describe 'what does that number means' without research, there are no tests about it (see https://github.com/DevExpress/DevExtreme/blob/21_1/testing/tests/DevExpress.ui.widgets/treeViewParts/events.js).
And I think it's not a good time to research WTF is that "Object" after 'codefreeze': there are no comments about it in initial PR, no tests and it's not easy to review all the possible scenarios and call stacks in code.

imho at the moment we can keep "readonly itemIndex?: number | any;"

File: js/events/index.d.ts (hunk 3 above)
Comment (IgnatovDan) [reply js/events/index.d.ts]: https://github.com/IlyaKhD/DevExtreme/pull/1/files : component/element/model members are readonly in event args

imho we can add 'readonly' here

Comment (DokaRus) [reply js/events/index.d.ts]: >imho we can add 'readonly' here

+1

File: js/events/index.d.ts (hunk 4 above)
Comment (DokaRus) [reply js/events/index.d.ts]: As a usual, for our users, the presence of this field is an indicator of whether the event is the result of end-user interaction, or the result of using the API

File: js/events/index.d.ts [hunk 7]
```diff
@@ -6,6 +10,50 @@ export interface EventType { }
...
+    component?: T;
+    element?: TElement;
+}
+
+/** @public */
+export interface ComponentOptionChangedEvent<T> {
+    component?: T;
+    name?: string;
+    fullName?: string;
+    value?: any;
```
Comment (DokaRus) [inline js/events/index.d.ts]: Is `previousValue` field missed?

File: js/events/index.d.ts [hunk 8]
```diff
@@ -6,6 +10,50 @@ export interface EventType { }
...
+export interface ComponentInitializedEvent<T> {
+    component?: T;
+    element?: TElement;
+}
+
+/** @public */
+export interface ComponentOptionChangedEvent<T> {
+    component?: T;
+    name?: string;
+    fullName?: string;
```
Comment (DokaRus) [inline js/events/index.d.ts]: Should we remove optional modifier for this fields?

File: js/events/index.d.ts (hunk 2 above)
Comment (DokaRus) [reply js/events/index.d.ts]: >imho we can add 'readonly' here

+1

File: js/ui/list.d.ts (hunk 1 above)
Comment (DokaRus) [reply js/ui/list.d.ts:87]: Yes, it looks inconsistent, but it reflects real behavior. `ItemRendered` in this field does not reflect whether the item belongs to a group.
I suggest we leave it as it is. I'll add the consistency discussion card for this event to our board and we'll fix it separately

File: js/ui/drop_down_button.d.ts [hunk 9]
```diff
@@ -30,6 +35,35 @@ import Widget, {
...
+/** @public */
+export type DisposingEvent = ComponentDisposingEvent<dxDropDownButton>;
+
+/** @public */
+export type InitializedEvent = ComponentInitializedEvent<dxDropDownButton>;
+
+/** @public */
+export type ItemClickEvent = ComponentNativeEvent<dxDropDownButton> & {
+    readonly itemData?: any;
+    itemElement: TElement;
```
Comment (DokaRus) [inline js/ui/drop_down_button.d.ts]:
```suggestion
    readonly itemElement: TElement;
```

File: js/ui/list.d.ts [hunk 10]
```diff
@@ -31,6 +37,87 @@ import {
     SearchBoxMixinOptions
 } from './widget/ui.search_box_mixin';
 
+interface ListItemInfo {
+    readonly itemData?: any;
+    itemElement: TElement;
```
Comment (DokaRus) [inline js/ui/list.d.ts]:
```suggestion
    readonly itemElement: TElement;
```

File: js/ui/list.d.ts [hunk 11]
```diff
@@ -31,6 +37,87 @@ import {
...
+/** @public */
+export type ContentReadyEvent = ComponentEvent<dxList>;
+
+/** @public */
+export type DisposingEvent = ComponentDisposingEvent<dxList>;
+
+/** @public */
+export type GroupRenderedEvent = ComponentEvent<dxList> & {
+    readonly groupData?: any;
+    groupElement?: TElement;
```
Comment (DokaRus) [inline js/ui/list.d.ts]:
```suggestion
    readonly groupElement?: TElement;
```

File: js/ui/popup.d.ts [hunk 12]
```diff
@@ -14,11 +14,64 @@ import {
...
     dxOverlayAnimation,
     dxOverlayOptions
 } from './overlay';
 
+import {
+    ResizeInfo
+} from './resizable';
+
+export interface TitleRenderedInfo {
+    titleElement: TElement
```
Comment (DokaRus) [inline js/ui/popup.d.ts]:
```suggestion
    readonly titleElement: TElement
```

**Review by DokaRus — COMMENTED** [review]
@groshenkovamarina  could you review SpeedDialAction changes?

File: js/events/index.d.ts (hunk 4 above)
Comment (IlyaKhD) [reply js/events/index.d.ts]: Ok, let's keep it optional for now

File: js/events/index.d.ts (hunk 8 above)
Comment (IlyaKhD) [reply js/events/index.d.ts]: done

File: js/ui/tree_view.d.ts (hunk 5 above)
Comment (IlyaKhD) [reply js/ui/tree_view.d.ts:45]: Ok

File: js/ui/form.d.ts (hunk 6 above)
Comment (IlyaKhD) [reply js/ui/form.d.ts:73]: Great, thank you

File: js/events/index.d.ts (hunk 7 above)
Comment (IlyaKhD) [reply js/events/index.d.ts]: Fixed, thanks

**Review by DokaRus — APPROVED** [review]

**Pull request merged**